source/server-discovery-and-monitoring/tests/README.rst

```diff
@@ -262,6 +262,38 @@ Run the following test(s) on MongoDB 4.4+.
              mode: "off",
          });
 
+Heartbeat Tests
+~~~~~~~~~~~~~~~
+
+1. Test that ``ServerHeartbeatStartedEvent`` is emitted after the monitoring socket was created and before the ``hello`` is sent to the server
+
+   #. Create a mock TCP server (example shown below) that pushes a ``client connected`` event to a shared array when a client connects and a ``client hello received`` event when the server receives the client hello and then closes the connection::
+         
+        let events = [];
+        server = createServer(clientSocket => {
+          events.push('client connected');
+
+          clientSocket.on('data', () => {
+            events.push('client hello received');
+            clientSocket.end();
+          });
+        });
+        server.listen(9999);
+
+   #. Create a client with ``serverSelectionTimeoutMS: 500, maxPoolSize: 1, minPoolSize: 0`` and listen to ``ServerHeartbeatStartedEvent`` and ``ServerHeartbeatFailedEvent``, pushing the event name to the same shared array as the mock TCP server.
+
+   #. Attempt to connect client to previously created TCP server, catching the error when the client fails to connect 
+
+   #. Assert that the event array has a length of 4
+
+   #. Assert that the first event is the ``client connected`` event
+
+   #. Assert that the second event is the ``serverHeartbeatStartedEvent``
+
+   #. Assert that the third event is the ``client hello received`` event
+
+   #. Assert that the fourth event is the ``serverHeartbeatFailedEvent``
+
```

Comment:
Do we think this test will be accurate with respect to the relative ordering of events on the driver side even though we'll be using sockets (albeit locally)?

This seems to be the best approach for now since we don't have more granular logging for when the the monitoring connection is created and when the hello itself is sent.

Reply:
Seems so to me. 

> even though we'll be using sockets

What is the implied concern here?

Reply:
I was thinking of whether or not the delay between creating the connection and emitting the `ServerHeartbeatStartedEvent` on the client side would be shorter than the time it takes for the mock server to emit the `client connected` event as in the following sequence diagram:

|client|server|
|------|------|
| create connection | . |
| emit heartbeat started| . |
| . | receive connection; emit `client connected` |
| send hello | . |
| . | receive hello; emit `client hello received`; close connection |
| emit heartbeat failed | . |

which would lead the `events` buffer to look like this

```typescript
[
	'serverHeartbeatStartedEvent',
	'client connected',
    'client hello received',
    'serverHeartbeatFiledEvent'
]
```
and report a false failure since the client actually performed the steps in the correct order

Reply:
Makes sense. There is a bit of an assumption going on here that there wouldn't be any multithreading or multiprocessing forcing the server's "client connected" to strictly proceed the driver code getting an opportunity to run whatever "connection established" handling it has. 

Perhaps the test could divest its core assertions a bit from the suggested implementation. A compiled language may be able to accomplish the same goal if it were to supplant the socket API with a function under its control to get notified of socket construction. 

I hope there is enough prior art to make this test accessible:

- https://nodejs.org/docs/latest/api/net.html#class-netserver
- https://docs.python.org/3/library/socketserver.html#socketserver-tcpserver-example
- https://ruby-doc.org/stdlib-2.4.0/libdoc/socket/rdoc/TCPServer.html
- https://docs.oracle.com/javase/8/docs/api/java/net/ServerSocket.html
- https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.tcplistener?view=net-8.0

Reply:
Update: we want the order to be 

```
['serverHeartbeatStarted', 'client connected', 'client hello received', 'serverHeartbeatFailed']
```